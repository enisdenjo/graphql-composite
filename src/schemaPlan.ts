import {
  ConstDirectiveNode,
  GraphQLField,
  GraphQLObjectType,
  GraphQLSchema,
  Kind,
} from 'graphql';

export interface SchemaPlan {
  types: Record<string, SchemaPlanType>;
}

export interface SchemaPlanType {
  sources: [SchemaPlanSource, SchemaPlanResolver][];
  fields: SchemaPlanField[];
}

export interface SchemaPlanField {
  sources: [SchemaPlanSource, SchemaPlanResolver | null][];
  name: string;
}

export interface SchemaPlanSource {
  /** Unique subgraph name. */
  subgraph: string;
  /** Name of the type or field in the source subgraph. */
  name: string;
}

export interface SchemaPlanResolver {
  operation: string;
  kind: 'FETCH';
  variables: Variable[];
}

interface Variable {
  name: string;
  /** Which field in the type to use as this variable. */
  select: string;
}

export function planSchema(schema: GraphQLSchema): SchemaPlan {
  const plan: SchemaPlan = { types: {} };

  const types = schema.getTypeMap();

  for (const [name, type] of Object.entries(types)) {
    if (name.startsWith('__')) {
      // ignore introspection types
      continue;
    }
    if (!(type instanceof GraphQLObjectType)) {
      // ignore non object types
      continue;
    }

    const planType: SchemaPlanType = {
      sources: [],
      fields: [],
    };

    for (const source of getSourcesFromDirectives(type, null)) {
      const resolver = getResolverForSourceFromDirectives(type, null, source);
      planType.sources.push([source, resolver]);
    }

    for (const [name, field] of Object.entries(type.getFields())) {
      const planField: SchemaPlanField = {
        sources: [],
        name,
      };

      for (const source of getSourcesFromDirectives(type, field)) {
        const resolver = getResolverForSourceFromDirectives(
          type,
          field,
          source,
        );
        planField.sources.push([source, resolver]);
      }

      planType.fields.push(planField);
    }

    plan.types[type.name] = planType;
  }

  return plan;
}

/**
 * Gets the sources of either the `field`, if provided, or the `type`.
 * Both arguments are used to better the error messages.
 */
function getSourcesFromDirectives(
  type: GraphQLObjectType,
  field: GraphQLField<unknown, unknown, unknown> | null,
): SchemaPlanSource[] {
  const fieldOrType = field || type;
  const sources: SchemaPlanSource[] = [];
  for (const sourceDirective of fieldOrType.astNode?.directives?.filter(
    (d) => d.name.value === 'source',
  ) || []) {
    // TODO: make sure source doesnt already exist
    sources.push({
      subgraph: mustGetStringArgumentValue(
        type,
        field,
        sourceDirective,
        'subgraph',
      ),
      name: mustGetStringArgumentValue(type, field, sourceDirective, 'name'),
    });
  }
  return sources;
}

/**
 * Gets the resolver of the source in either the `field`, if provided, or the `type`.
 * Both arguments are used to better the error messages.
 */
function getResolverForSourceFromDirectives(
  type: GraphQLObjectType,
  field: null,
  source: SchemaPlanSource,
): SchemaPlanResolver;
function getResolverForSourceFromDirectives(
  type: GraphQLObjectType,
  field: GraphQLField<unknown, unknown, unknown>,
  source: SchemaPlanSource,
): SchemaPlanResolver | null;
function getResolverForSourceFromDirectives(
  type: GraphQLObjectType,
  field: GraphQLField<unknown, unknown, unknown> | null,
  source: SchemaPlanSource,
): SchemaPlanResolver | null {
  const fieldOrType = field || type;
  const directives = fieldOrType.astNode?.directives || [];
  const resolverDirs = directives.filter(
    (d) =>
      d.name.value === 'resolver' &&
      mustGetStringArgumentValue(type, field, d, 'subgraph') ===
        source.subgraph,
  );
  const resolverDir = resolverDirs[0];
  if (!resolverDir) {
    if (field === fieldOrType) {
      // sources on fields must not always have a resolver
      return null;
    }
    throw new Error(
      `source "${source.subgraph}" on type "${type.name}" has no resolver directive`,
    );
  }
  if (resolverDirs.length > 1) {
    throw new Error(
      `source "${source.subgraph}" on type "${type.name}" has multiple resolver directives`,
    );
  }

  const variables: Variable[] = [];
  for (const variableDir of directives.filter(
    (d) =>
      d.name.value === 'variable' &&
      mustGetStringArgumentValue(type, field, d, 'subgraph') ===
        source.subgraph,
  )) {
    variables.push({
      name: mustGetStringArgumentValue(type, field, variableDir, 'name'),
      select: mustGetStringArgumentValue(type, field, variableDir, 'select'),
    });
  }

  return {
    kind: 'FETCH', // TODO: other kinds
    operation: mustGetStringArgumentValue(
      type,
      field,
      resolverDir,
      'operation',
    ),
    variables,
  };
}

function mustGetStringArgumentValue(
  type: GraphQLObjectType,
  field: GraphQLField<unknown, unknown, unknown> | null,
  directive: ConstDirectiveNode,
  name: string,
): string {
  const arg = directive.arguments?.find((s) => s.name.value === name)?.value;
  if (!arg) {
    throw new Error(
      `directive @${directive.name.value} on type "${type.name}"${field ? ` at field "${field.name}"` : ''} doesnt have an "${name}" argument`,
    );
  }
  if (arg.kind !== Kind.STRING) {
    throw new Error(
      `directive @${directive.name.value} on type "${type.name}"${field ? ` at field "${field.name}"` : ''} argument "${name}" is not a string`,
    );
  }
  return arg.value;
}
