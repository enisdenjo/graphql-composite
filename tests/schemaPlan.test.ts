import fs from 'fs/promises';
import path from 'path';
import { buildSchema } from 'graphql';
import { expect, it } from 'vitest';
import { planSchema } from '../src/schemaPlan.js';

it('should plan schema', async () => {
  const supergraph = await fs.readFile(
    path.resolve(__dirname, 'fixtures', 'fusiongraph.graphql'),
    'utf-8',
  );
  expect(
    planSchema(
      buildSchema(supergraph, {
        assumeValid: true,
      }),
    ),
  ).toMatchSnapshot();
});
