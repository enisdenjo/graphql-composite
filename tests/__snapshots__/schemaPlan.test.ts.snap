// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should plan schema 1`] = `
{
  "types": {
    "Manufacturer": {
      "fields": [
        {
          "name": "id",
          "sources": [
            [
              {
                "name": "id",
                "subgraph": "manufacturers",
              },
              null,
            ],
            [
              {
                "name": "id",
                "subgraph": "products",
              },
              null,
            ],
          ],
        },
        {
          "name": "name",
          "sources": [
            [
              {
                "name": "name",
                "subgraph": "manufacturers",
              },
              null,
            ],
          ],
        },
        {
          "name": "products",
          "sources": [
            [
              {
                "name": "products",
                "subgraph": "products",
              },
              null,
            ],
          ],
        },
      ],
      "sources": [
        [
          {
            "name": "Manufacturer",
            "subgraph": "manufacturers",
          },
          {
            "kind": "FETCH",
            "operation": "query ManufacturerById($Manufacturer_id: ID!) { manufacturer(id: $Manufacturer_id) }",
            "variables": [
              {
                "name": "Manufacturer_id",
                "select": "id",
              },
              {
                "name": "Manufacturer_id",
                "select": "id",
              },
            ],
          },
        ],
        [
          {
            "name": "Manufacturer",
            "subgraph": "products",
          },
          {
            "kind": "FETCH",
            "operation": "query ManufacturerById($Manufacturer_id: ID!) { _manufacturer(id: $Manufacturer_id) }",
            "variables": [
              {
                "name": "Manufacturer_id",
                "select": "id",
              },
              {
                "name": "Manufacturer_id",
                "select": "id",
              },
            ],
          },
        ],
      ],
    },
    "Product": {
      "fields": [
        {
          "name": "manufacturer",
          "sources": [
            [
              {
                "name": "manufacturer",
                "subgraph": "products",
              },
              null,
            ],
          ],
        },
        {
          "name": "name",
          "sources": [
            [
              {
                "name": "name",
                "subgraph": "products",
              },
              null,
            ],
          ],
        },
        {
          "name": "price",
          "sources": [
            [
              {
                "name": "price",
                "subgraph": "products",
              },
              null,
            ],
          ],
        },
        {
          "name": "upc",
          "sources": [
            [
              {
                "name": "upc",
                "subgraph": "products",
              },
              null,
            ],
            [
              {
                "name": "upc",
                "subgraph": "storefronts",
              },
              null,
            ],
          ],
        },
      ],
      "sources": [
        [
          {
            "name": "Product",
            "subgraph": "products",
          },
          {
            "kind": "FETCH",
            "operation": "query ProductByUpc($Product_upc: ID!) { product(upc: $Product_upc) }",
            "variables": [
              {
                "name": "Product_upc",
                "select": "upc",
              },
            ],
          },
        ],
        [
          {
            "name": "Product",
            "subgraph": "storefronts",
          },
          {
            "kind": "FETCH",
            "operation": "TODO",
            "variables": [
              {
                "name": "Product_upc",
                "select": "upc",
              },
            ],
          },
        ],
      ],
    },
    "Query": {
      "fields": [
        {
          "name": "manufacturer",
          "sources": [
            [
              {
                "name": "manufacturer",
                "subgraph": "manufacturers",
              },
              {
                "kind": "FETCH",
                "operation": "query manufacturer($id: ID!) { manufacturer(id: $id) }",
                "variables": [],
              },
            ],
          ],
        },
        {
          "name": "product",
          "sources": [
            [
              {
                "name": "product",
                "subgraph": "products",
              },
              {
                "kind": "FETCH",
                "operation": "query product($upc: ID!) { product(upc: $upc) }",
                "variables": [],
              },
            ],
          ],
        },
        {
          "name": "_manufacturer",
          "sources": [
            [
              {
                "name": "_manufacturer",
                "subgraph": "products",
              },
              {
                "kind": "FETCH",
                "operation": "query _manufacturer($id: ID!) { _manufacturer(id: $id) }",
                "variables": [],
              },
            ],
          ],
        },
        {
          "name": "storefront",
          "sources": [
            [
              {
                "name": "storefront",
                "subgraph": "storefronts",
              },
              {
                "kind": "FETCH",
                "operation": "query storefront($id: ID!) { storefront(id: $id) }",
                "variables": [],
              },
            ],
          ],
        },
      ],
      "sources": [],
    },
    "Storefront": {
      "fields": [
        {
          "name": "id",
          "sources": [
            [
              {
                "name": "id",
                "subgraph": "storefronts",
              },
              null,
            ],
          ],
        },
        {
          "name": "name",
          "sources": [
            [
              {
                "name": "name",
                "subgraph": "storefronts",
              },
              null,
            ],
          ],
        },
        {
          "name": "products",
          "sources": [
            [
              {
                "name": "products",
                "subgraph": "storefronts",
              },
              null,
            ],
          ],
        },
      ],
      "sources": [
        [
          {
            "name": "Storefront",
            "subgraph": "storefronts",
          },
          {
            "kind": "FETCH",
            "operation": "query StorefrontById($Storefront_id: ID!) { storefront(id: $Storefront_id) }",
            "variables": [
              {
                "name": "Storefront_id",
                "select": "id",
              },
            ],
          },
        ],
      ],
    },
  },
}
`;
